### PR TITLE
Inliner method resolution keeps concrete methods

### LINKED ISSUE #12787: Scala 2.13 inline optimizer unable to correctly look up `.head` method in `scala.collection.LinearSeq`
## Reproduction steps

The current Scala 2.12/2.13 optimizer appears to be unable to correctly look up the `.head` method from the Scala `scala.collection.LinearSeq` when you happen to use `@inline` on a method that ends up transitively calling `.head`. The context of this is the current PR at https://github.com/apache/incubator-pekko/pull/305. If you checkout the branch and run `sbt ++2.13.10 actor/compile` you will get the following error messages

```
[error] /Users/mdedetrich/github/incubator-pekko/actor/src/main/scala-2.13/org/apache/pekko/util/ByteIterator.scala:370:50: org/apache/pekko/util/ByteIterator$MultiByteArrayIterator::org$apache$pekko$util$ByteIterator$MultiByteArrayIterator$$current()Lorg/apache/pekko/util/ByteIterator$ByteArrayIterator; is annotated @inline but could not be inlined:
[error] Failed to check if org/apache/pekko/util/ByteIterator$MultiByteArrayIterator::org$apache$pekko$util$ByteIterator$MultiByteArrayIterator$$current()Lorg/apache/pekko/util/ByteIterator$ByteArrayIterator; can be safely inlined to org/apache/pekko/util/ByteIterator$MultiByteArrayIterator without causing an IllegalAccessError. Checking instruction INVOKEINTERFACE scala/collection/LinearSeq.head ()Ljava/lang/Object; (itf) failed:
[error] The method head()Ljava/lang/Object; could not be found in the class scala/collection/LinearSeq or any of its parents.
[error]       getToArray(xs, offset, n, 1) { getByte } { current.getBytes(_, _, _) }
[error]                                                  ^
[error] /Users/mdedetrich/github/incubator-pekko/actor/src/main/scala-2.13/org/apache/pekko/util/ByteIterator.scala:385:63: org/apache/pekko/util/ByteIterator$MultiByteArrayIterator::org$apache$pekko$util$ByteIterator$MultiByteArrayIterator$$current()Lorg/apache/pekko/util/ByteIterator$ByteArrayIterator; is annotated @inline but could not be inlined:
[error] Failed to check if org/apache/pekko/util/ByteIterator$MultiByteArrayIterator::org$apache$pekko$util$ByteIterator$MultiByteArrayIterator$$current()Lorg/apache/pekko/util/ByteIterator$ByteArrayIterator; can be safely inlined to org/apache/pekko/util/ByteIterator$MultiByteArrayIterator without causing an IllegalAccessError. Checking instruction INVOKEINTERFACE scala/collection/LinearSeq.head ()Ljava/lang/Object; (itf) failed:
[error] The method head()Ljava/lang/Object; could not be found in the class scala/collection/LinearSeq or any of its parents.
[error]       getToArray(xs, offset, n, 8) { getDouble(byteOrder) } { current.getDoubles(_, _, _)(byteOrder) }
...
```

The `current` method is defined as 
```scala
@inline private def current: ByteArrayIterator = iterators.head
```

More context can be found at https://contributors.scala-lang.org/t/scala-2-12-2-13-inliner-is-too-aggressive/6191. Will also try to create a minimised example

## Problem

It appears that specifically for Scala 2.13, the optimizer has issues in inlining methods that happen to call `.head` from `scala.collection.LinearSeq` in the call chain. I suspect that this may be a result of the new collections included in Scala 2.13.

### PR DESCRIPTION
**Pull request by som-snytt**
IterableOps.head was knocked out by abstract LinearSeqOps.head.

Fixes scala/bug#12787

### CODE REVIEW

File: src/compiler/scala/tools/nsc/backend/jvm/opt/ByteCodeRepository.scala [hunk 1]
```diff
@@ -217,19 +217,19 @@ abstract class ByteCodeRepository extends PerRunInit {
       val result =
         if (found.size <= 1) found.headOption
         else {
-          val maxSpecific = found.filterNot({
+          val maxSpecific = found.filterNot {
             case (method, owner) =>
               isAbstractMethod(method) || {
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/opt/ByteCodeRepository.scala]: This also seems to be off-spec (see below).

File: src/compiler/scala/tools/nsc/backend/jvm/opt/ByteCodeRepository.scala [hunk 2]
```diff
@@ -217,19 +217,19 @@ abstract class ByteCodeRepository extends PerRunInit {
...
                 val ownerTp = bTypesFromClassfile.classBTypeFromClassNode(owner)
-                found exists {
+                found.exists {
                   case (other, otherOwner) =>
-                    (other ne method) && {
+                    (other ne method) && !isAbstractMethod(other) && {
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/opt/ByteCodeRepository.scala]: I think this is not according to the spec (https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-5.html#jvms-5.4.3.3)

>  Where the method is declared in interface I, there exists no other maximally-specific superinterface method of C with the specified name and descriptor that is declared in a subinterface of I

There's no mention of `abstract`.

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/backend/jvm/opt/ByteCodeRepository.scala]: Under bullet item 3,
> exactly one method that does not have its ACC_ABSTRACT flag set

Funny enough, I was reading the latest spec and started opening up this code, then decided that was too much and tried to do just the minimal delta. I'll try to refresh and reload again, it did not stick with me.

Comment (lrytz) [reply src/compiler/scala/tools/nsc/backend/jvm/opt/ByteCodeRepository.scala]: To expand the quote a bit

>  If the maximally-specific superinterface methods of C for the name and descriptor specified by the method reference include exactly one method that does not have its ACC_ABSTRACT flag set

So filtering abstract is after getting the "maximally-specific" methods, which is where we are here.

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/backend/jvm/opt/ByteCodeRepository.scala]: OK I see what you mean.

**Comment by som-snytt** [comment]
@lrytz just let me know if the refactor is annoying or not clarifying. I always wanted to use `Either`.

**Comment by lrytz** [comment]
LGTM, no more non-local return, though the old version was slightly more straigthforward to follow

**Comment by som-snytt** [comment]
I guess you mean the old version was more work to follow straight. :)

I'll give it another look for an optimal balance of naming things and returning early from them.

**Comment by som-snytt** [comment]
Since the local refactor did not receive a ringing endorsement (wherever that idiom came from), I rebased with just the `wip`. There was a conflict on rebase dropping the first commit, and I just took the `wip` version.

I found the combinators easier to read, but I did not consider performance. Probably it only matters if you can read it.

Also, feel free to squash or open a fresh PR.

**Review by lrytz — APPROVED** [review]
LGTM, merge after 2.13.11

**Comment by mdedetrich** [comment]
Would it be possible to merge this now that Scala 2.13.11 has been released? Would like to try a nightly release with this fix just to triple check everything works as expected.

**Pull request merged**